react/AppIcon/index.jsx

```diff
@@ -0,0 +1,88 @@
+import React, { Component } from 'react'
+import PropTypes from 'prop-types'
+import classNames from 'classnames'
+import styles from './styles.styl'
+
+import Icon from '../Icon'
+
+import appDefaultIcon from '../../assets/icons/ui/icon-cube.svg'
+
+export class AppIcon extends Component {
+  constructor(props, context) {
+    super(props, context)
+    this.state = {
+      status: 'idle'
+    }
+  }
+
+  async componentDidMount() {
+    const { app } = this.props
+    const { client } = this.context
+    const cozyClient = client && client.link && client.link.client
+    if (!cozyClient || typeof cozyClient.fetch !== 'function') {
+      console.warn(
+        'AppIcon needs CozyClient and to be called into a <CozyClientProvider /> component.'
+      )
+      this.setState({ status: 'failed' })
+      return
+    }
+    const icon = await this.fetchIcon(app.links.icon, cozyClient)
+    this.setState({ icon, status: icon ? 'done' : 'failed' })
```

Comment:
IMHO, all this should be done in fetchIcon. It should not be the responsibility of `componentDidMount` to do that. 

```
componentDidMount() {
  this.fetchIcon()
}
```

This let the fetchIcon be the only responsible for updating the `status` state property.